### PR TITLE
feat: Improve perf of AllGather-Top1 after LMHead

### PR DESCRIPTION
**Pull request by nvpohanh**
[feat: Improve perf of AllGather-Top1 after LMHead](https://github.com/NVIDIA/TensorRT-LLM/commit/5591b8d9151df5c1d0cdc50a76d94c499d0e2c52) 

Currently, after we run LMHead, we first run AllGather to collect the
logits for the entire vocab size, and then apply top1. This has two perf
issues: (1) AllGather is slow (2) we are doing redundant top1 across tp
ranks.

Therefore, replace the AllGather-Top1 with Top1-AllReduce-Top1. The idea
is to first get the per-rank top1, and then use AllReduce to collect the
per-rank top1 argmax/max values, and then do argmax again.

This change requires turning off the AllGather in LMHead as well as
handling the additional AllReduce in Eagle3OneModelWorker.

### CODE REVIEW

**Comment by nvpohanh** [comment]
```
[0] Prompt: 'Hello, my name is', Generated text: ' {{ name }} and I am {{ age }} years old.\n    </body>\n</html>\n"""\n\n# Define the data to be used in the template\ndata = {\n    "title": "My Page",\n    "name": "John Doe",\n    "age": 30\n}\n\n# Render the template with the'
[1] Prompt: 'The president of the United States is', Generated text: ' the head of state and head of government of the United States. The president leads the executive branch of the federal government and is the commander-in-chief of the United States Armed Forces. The president is responsible for executing and enforcing the laws created by Congress. The president is also responsible for appointing federal judges, ambassadors'
[2] Prompt: 'The capital of France is', Generated text: ' Paris, and it is known for its rich history, art, and culture. The city is often referred to as the "City of Light" due to its role in the Enlightenment and its many famous landmarks, such as the Eiffel Tower and the Louvre Museum. Paris is a global center for fashion, cuisine'
[3] Prompt: 'The future of AI is', Generated text: ' a topic of much debate and speculation. As we continue to develop and refine AI technologies, what can we expect the future of AI to hold, and how will it impact various aspects of our lives and society as a whole?\n\n1. **Advancements in AI Capabilities**: The future of AI is expected to be marked by'
```

**Comment by nvpohanh** [comment]
MMLU:

```
MMLU weighted average accuracy: 86.07 (14042)
```

**Comment by nvpohanh** [comment]
GPQA: 

```
Writing report to /scratch/results/gpqa_diamond.html                                                                                                                                                          
{'score:std': np.float64(0.49936190280033443), 'score:stderr': 0.035578062450873145, 'score': np.float64(0.5252525252525253), 'task_name': 'gpqa_diamond'}
```
